src/libstd/ffi/c_str.rs

```diff
@@ -197,7 +199,35 @@ impl CString {
     #[stable(feature = "rust1", since = "1.0.0")]
     pub unsafe fn from_vec_unchecked(mut v: Vec<u8>) -> CString {
         v.push(0);
-        CString { inner: v }
+        CString { inner: v.into_boxed_slice() }
+    }
+
+    /// Retakes ownership of a CString that was transferred to C.
+    ///
+    /// The only appropriate argument is a pointer obtained by calling
+    /// `into_ptr`.
+    #[unstable(feature = "c_str_memory", reason = "recently added")]
+    pub unsafe fn from_ptr(ptr: *const libc::c_char) -> CString {
+        let len = libc::strlen(ptr);
+        let len_with_nul = len as usize + 1;
+        let vec = Vec::from_raw_parts(ptr as *mut u8, len_with_nul, len_with_nul);
```

Comment:
Instead of going through `Vec`, I think it may be best to go directly to a `Box<[u8]>` because it's where the pointer originally came from, for example:

````rust
let len = libc:;strlen(ptr) + 1;
let slice = slice::from_raw_parts(ptr, len as usize);
CString { inner: mem::transmute(slice) }
````

Going through `Vec` and then possibly hitting `shrink_to_fit` may be adding a bit more indirection than necessary (and may also be somewhat less robust)